From 708ea520fac61d30952e614860165d37628f012b Mon Sep 17 00:00:00 2001
From: Tanish2002 <tanishkhare@gmail.com>
Date: Tue, 15 Dec 2020 15:10:18 +0530
Subject: [PATCH 09/11] SIGUSR1 Reload

---
 x.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/x.c b/x.c
index 93444f8..021099d 100644
--- a/x.c
+++ b/x.c
@@ -2346,6 +2346,28 @@ usage(void)
 	    " [stty_args ...]\n", argv0, argv0);
 }
 
+void
+reload(int sig)
+{
+	config_init();
+
+	/* colors, fonts */
+	xloadcols();
+	xunloadfonts();
+	usedfont = (opt_font == NULL)? font : opt_font;
+	xloadfonts(usedfont, 0);
+
+	/* pretend the window just got resized */
+	cresize(win.w, win.h);
+
+	redraw();
+
+	/* triggers re-render if we're visible. */
+	ttywrite("\033[O", 3, 1);
+
+	signal(SIGUSR1, reload);
+}
+
 int
 main(int argc, char *argv[])
 {
@@ -2413,6 +2435,7 @@ run:
 	config_init();
 	cols = MAX(cols, 1);
 	rows = MAX(rows, 1);
+	signal(SIGUSR1, reload);
 	tnew(cols, rows);
 	xinit(cols, rows);
 	xsetenv();
-- 
2.29.2

