From be8d445ffc8f5989c558cb26b449b6bda765899e Mon Sep 17 00:00:00 2001
From: Tanish2002 <tanishkhare@gmail.com>
Date: Tue, 15 Dec 2020 15:11:48 +0530
Subject: [PATCH 11/12] st-w3m

---
 x.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/x.c b/x.c
index 619961e..3855f13 100644
--- a/x.c
+++ b/x.c
@@ -1827,6 +1827,8 @@ xsixelscrollup(SixelContext *ctx, int n, int top)
 int
 xstartdraw(void)
 {
+	if (IS_SET(MODE_VISIBLE))
+		XCopyArea(xw.dpy, xw.win, xw.buf, dc.gc, 0, 0, win.w, win.h, 0, 0);
 	return IS_SET(MODE_VISIBLE);
 }
 
-- 
2.29.2

