From a3222820334686c3fc70382eb0c5badf8c3377c2 Mon Sep 17 00:00:00 2001
From: Tanish2002 <tanishkhare@gmail.com>
Date: Sun, 13 Dec 2020 16:35:24 +0530
Subject: [PATCH 2/3] add w3m support

---
 x.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/x.c b/x.c
index dd9e628..20b1177 100644
--- a/x.c
+++ b/x.c
@@ -1768,6 +1768,8 @@ xsettitle(char *p)
 int
 xstartdraw(void)
 {
+	if (IS_SET(MODE_VISIBLE))
+		XCopyArea(xw.dpy, xw.win, xw.buf, dc.gc, 0, 0, win.w, win.h, 0, 0);
 	return IS_SET(MODE_VISIBLE);
 }
 
-- 
2.29.2

